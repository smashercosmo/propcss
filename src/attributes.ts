export const attributes: { [key: string]: string } = {
  m: 'margin',
  mt: 'margin-top',
  mb: 'margin-bottom',
  ml: 'margin-left',
  mr: 'margin-right',
  p: 'padding',
  pt: 'padding-top',
  pb: 'padding-bottom',
  pl: 'padding-left',
  pr: 'padding-right',
  w: 'width',
  h: 'height',
}