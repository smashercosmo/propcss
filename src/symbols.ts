export const PLUGIN_NAMESPACE = Symbol('propcss')
