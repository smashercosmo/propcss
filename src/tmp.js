import { Box } from 'propcss/components'
import React from 'react'

export function Test2() {
  return (
    <Box h="20px" pl="40px">Hello</Box>
  )
}