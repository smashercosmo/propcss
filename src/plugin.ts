/* eslint-disable no-param-reassign */
import { types as t, NodePath, PluginObj, BabelFileMetadata } from '@babel/core'
import htmlTags from 'html-tags'

import { PLUGIN_NAMESPACE } from './symbols'
import { normalizeClassName } from './utils'
import { PluginOptions } from './types'

type File = {
  metadata: BabelFileMetadata & {
    [PLUGIN_NAMESPACE]?: { [attr: string]: Set<string> }
  }
  set(key: any, val: any): void
  get(key: any): any
  has(key: any): boolean
}

type State = {
  file: File
  opts: PluginOptions
}

function isAttrAllowed(aliasMap: { [key: string]: string }, attr: string) {
  return Boolean(aliasMap[attr])
}

function getAttributeValue(attribute: t.JSXAttribute | t.JSXSpreadAttribute) {
  if (!t.isJSXAttribute(attribute)) {
    return undefined
  }

  if (t.isStringLiteral(attribute.value)) {
    return attribute.value
  }

  if (
    t.isJSXExpressionContainer(attribute.value) &&
    !t.isJSXElement(attribute.value.expression) &&
    !t.isJSXFragment(attribute.value.expression) &&
    !t.isJSXEmptyExpression(attribute.value.expression)
  ) {
    return attribute.value.expression
  }

  return undefined
}

/**
 * Generates jsx expression for className attribute:
 * "className={["atomicClass1 atomicClass2", originalClassNameAttributeExpression].join(' ')}"
 */
function generateClassName(
  classes: Set<string | number>,
  classNameAttribute?: t.JSXAttribute,
) {
  let classNameAttributeValue = classNameAttribute
    ? getAttributeValue(classNameAttribute)
    : undefined

  if (classes.size === 0) {
    return classNameAttribute
  }

  return t.jsxAttribute(
    t.jsxIdentifier('className'),
    t.jsxExpressionContainer(
      t.callExpression(
        t.memberExpression(
          t.arrayExpression([
            t.stringLiteral(Array.from(classes).join(' ')),
            classNameAttributeValue || null,
          ]),
          t.identifier('join'),
        ),
        [t.stringLiteral(' ')],
      ),
    ),
  )
}

function handleStringLiteral(value: t.StringLiteral, values: Set<string>) {
  values.add(value.value)
}

function handleNumericLiteral(value: t.NumericLiteral, values: Set<string>) {
  values.add(String(value.value))
}

function handleUnaryExpression(
  expression: t.UnaryExpression,
  values: Set<string>,
) {
  if (expression.operator === '-' && t.isNumericLiteral(expression.argument)) {
    values.add(expression.operator + String(expression.argument.value))
  }
}

function handleExpressionContainer(
  value: t.JSXExpressionContainer,
  values: Set<string>,
) {
  if (t.isNumericLiteral(value.expression)) {
    handleNumericLiteral(value.expression, values)
  } else if (t.isUnaryExpression(value.expression)) {
    handleUnaryExpression(value.expression, values)
  }
}

export function plugin(_api: any, options: PluginOptions): PluginObj<State> {
  let {
    componentPropToCSSPropMapping,
    CSSPropToClassNameMapping,
    components,
  } = options

  let allowedComponents = components ? new Set(components) : new Set(htmlTags)

  return {
    pre(file: File) {
      if (!file.has(PLUGIN_NAMESPACE)) {
        file.set(PLUGIN_NAMESPACE, {})
      }
    },
    post(file: File) {
      file.metadata[PLUGIN_NAMESPACE] = file.get(PLUGIN_NAMESPACE)
    },
    visitor: {
      JSXOpeningElement(path: NodePath<t.JSXOpeningElement>, state) {
        let classes = new Set<string>()
        let classNameAttribute: t.JSXAttribute | undefined
        let nonAtomicAttributes = []

        if (
          t.isJSXIdentifier(path.node.name) &&
          allowedComponents.has(path.node.name.name)
        ) {
          for (let i = 0, l = path.node.attributes.length; i < l; i += 1) {
            let attribute = path.node.attributes[i]
            if (
              t.isJSXAttribute(attribute) &&
              t.isJSXIdentifier(attribute.name)
            ) {
              if (attribute.name.name === 'className') {
                classNameAttribute = attribute
              } else if (
                !isAttrAllowed(
                  componentPropToCSSPropMapping,
                  attribute.name.name,
                )
              ) {
                nonAtomicAttributes.push(attribute)
              } else {
                let { name: propName } = attribute.name
                let values: Set<string> = new Set()

                if (t.isStringLiteral(attribute.value)) {
                  handleStringLiteral(attribute.value, values)
                } else if (t.isJSXExpressionContainer(attribute.value)) {
                  handleExpressionContainer(attribute.value, values)
                }

                if (!state.file.get(PLUGIN_NAMESPACE)[propName]) {
                  state.file.get(PLUGIN_NAMESPACE)[propName] = new Set()
                }

                let valuesArr = Array.from(values)

                for (let ii = 0, ll = valuesArr.length; ii < ll; ii += 1) {
                  let value = valuesArr[ii]
                  let cssProp = componentPropToCSSPropMapping[propName]
                  classes.add(
                    normalizeClassName(
                      CSSPropToClassNameMapping[cssProp],
                      value,
                      false,
                    ),
                  )
                  state.file.get(PLUGIN_NAMESPACE)[propName].add(value)
                }
              }
            }
          }

          path.node.attributes = nonAtomicAttributes

          let newClassNameAttribute = generateClassName(
            classes,
            classNameAttribute,
          )

          if (newClassNameAttribute) {
            path.node.attributes.push(newClassNameAttribute)
          }
        }
      },
    },
  }
}
