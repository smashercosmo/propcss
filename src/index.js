import 'propcss/propcss.css'
import { Box } from 'propcss/components'
import React from 'react'

import { Test2 } from './tmp'

export function Test() {
  return (
    <Box m="20px" pt="40px"><Test2 /></Box>
  )
}