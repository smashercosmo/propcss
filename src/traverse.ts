import path from 'path'
import { transformSync, BabelFileResult, BabelFileMetadata } from '@babel/core'

import { plugin } from './plugin'
import { PLUGIN_NAMESPACE } from './symbols'

type Result =
  | (Omit<BabelFileResult, 'metadata'> & {
      metadata: BabelFileMetadata & {
        [PLUGIN_NAMESPACE]?: { [attr: string]: Set<string> }
      }
    })
  | null

export function traverse(
  source: string,
  filename: string,
  options: { component: string; attributes: { [key: string]: string } },
): Result {
  const extname = path.extname(filename);

  // This was copied directly from
  // https://github.com/4Catalyzer/astroturf/blob/master/src/traverse.js
  return transformSync(source, {
    filename,
    babelrc: false,
    code: true,
    ast: false,
    plugins: [[plugin, options]],
    parserOpts: {
      allowImportExportEverywhere: true,
      allowReturnOutsideFunction: true,
      allowSuperOutsideMethod: true,
      sourceType: 'unambiguous',
      plugins: [
        'jsx',
        extname === '.ts' || extname === '.tsx' ? 'typescript' : 'flow',
        'doExpressions',
        'objectRestSpread',
        ['decorators', { decoratorsBeforeExport: true }],
        'classProperties',
        'classPrivateProperties',
        'classPrivateMethods',
        'exportDefaultFrom',
        'exportNamespaceFrom',
        'asyncGenerators',
        'functionBind',
        'functionSent',
        'dynamicImport',
        'numericSeparator',
        'optionalChaining',
        'importMeta',
        'bigInt',
        'optionalCatchBinding',
        'throwExpressions',
        ['pipelineOperator', { proposal: 'minimal' }],
        'nullishCoalescingOperator',
      ],
    },
  }) as Result
}
