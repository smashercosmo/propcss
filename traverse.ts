import path from 'path';
import { transformSync } from '@babel/core';
import plugin from './babel-plugin';

export default function traverse(source: string, filename: string) {
  const extname = path.extname(filename);

  return transformSync(source, {
    filename,
    babelrc: false,
    code: false,
    ast: false,
    plugins: [[plugin]],
    parserOpts: {
      allowImportExportEverywhere: true,
      allowReturnOutsideFunction: true,
      allowSuperOutsideMethod: true,
      sourceType: 'unambiguous',
      plugins: [
        'jsx',
        extname === '.ts' || extname === '.tsx' ? 'typescript' : 'flow',
        'doExpressions',
        'objectRestSpread',
        ['decorators', { decoratorsBeforeExport: true }],
        'classProperties',
        'classPrivateProperties',
        'classPrivateMethods',
        'exportDefaultFrom',
        'exportNamespaceFrom',
        'asyncGenerators',
        'functionBind',
        'functionSent',
        'dynamicImport',
        'numericSeparator',
        'optionalChaining',
        'importMeta',
        'bigInt',
        'optionalCatchBinding',
        'throwExpressions',
        ['pipelineOperator', { proposal: 'minimal' }],
        'nullishCoalescingOperator',
      ],
    },
  });
}