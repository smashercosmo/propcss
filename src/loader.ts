import path from 'path'
import fs from 'fs'
import webpack from 'webpack'
import { getOptions } from 'loader-utils'
import postcss from 'postcss'
import postcssJs from 'postcss-js'

import { traverse } from './traverse'
import {
  componentPropToCSSPropMapping as defaultComponentPropToCSSPropMapping,
  CSSPropToClassNameMapping as defaultCSSPropToClassNameMapping,
} from './attributes'
import { PLUGIN_NAMESPACE } from './symbols'
import { normalizeClassName, normalizeValue } from './utils'
import { CommonOptions, LoaderOptions } from './types'

type PropCSSLoaderContext = webpack.loader.LoaderContext & {
  _compiler: webpack.Compiler & {
    [PLUGIN_NAMESPACE]?: {
      [resourcePath: string]: { [attr: string]: Set<string> }
    }
  }
}

function createClasses(
  propsChunks: ReadonlyArray<{ [attr: string]: Set<string> }>,
  options: CommonOptions,
) {
  let { componentPropToCSSPropMapping, CSSPropToClassNameMapping } = options
  let classes: {
    [className: string]: {
      [prop: string]: string
    }
  } = {}

  for (let l = propsChunks.length, i = 0; i < l; i += 1) {
    let propsChunk = propsChunks[i]
    let propsNames = Object.keys(propsChunk)

    for (let ll = propsNames.length, j = 0; j < ll; j += 1) {
      let propName = propsNames[j]

      if (componentPropToCSSPropMapping[propName]) {
        let propValues = Array.from(propsChunk[propName])

        for (let lll = propValues.length, k = 0; k < lll; k += 1) {
          let propValue = propValues[k]
          let cssProp = componentPropToCSSPropMapping[propName]
          let className = `.${normalizeClassName(
            CSSPropToClassNameMapping[cssProp],
            propValue,
            true,
          )}`

          classes[className] = {
            [cssProp]: normalizeValue(propValue),
          }
        }
      }
    }
  }

  return classes
}

function loader(
  this: PropCSSLoaderContext,
  source: string | Buffer,
): string | Buffer | void | undefined {
  let { _compiler: compiler } = this
  let options = getOptions(this) as LoaderOptions
  let {
    componentPropToCSSPropMapping = {},
    CSSPropToClassNameMapping = {},
    components,
    filename,
    path: filepath,
  } = options

  if (!compiler[PLUGIN_NAMESPACE]) {
    compiler[PLUGIN_NAMESPACE] = {}
  }

  if (typeof source === 'string') {
    let allComponentPropToCSSPropMappings = Object.assign(
      {},
      defaultComponentPropToCSSPropMapping,
      componentPropToCSSPropMapping,
    )
    let allCSSPropToClassNameMappings = Object.assign(
      {},
      defaultCSSPropToClassNameMapping,
      CSSPropToClassNameMapping,
    )
    let result = traverse(source, this.resourcePath, {
      components,
      componentPropToCSSPropMapping: allComponentPropToCSSPropMappings,
      CSSPropToClassNameMapping: allCSSPropToClassNameMappings,
    })

    if (
      result &&
      result.code &&
      result.metadata &&
      result.metadata[PLUGIN_NAMESPACE]
    ) {
      let props = result.metadata[PLUGIN_NAMESPACE]
      let cache = compiler[PLUGIN_NAMESPACE]

      if (cache && props) {
        cache[this.resourcePath] = props
        let classes = createClasses(Object.values(cache), {
          componentPropToCSSPropMapping: allComponentPropToCSSPropMappings,
          CSSPropToClassNameMapping: allCSSPropToClassNameMappings,
        })
        let file = path.join(filepath, filename)
        let { css } = postcss().process(classes, { parser: postcssJs })
        fs.writeFileSync(file, css, 'utf-8')
        return result.code
      }
    }
  }
  return source
}

// eslint-disable-next-line import/no-default-export
export default loader
