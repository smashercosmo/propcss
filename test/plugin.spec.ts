import { traverse } from '../src/traverse'
import {
  CSSPropToClassNameMapping,
  componentPropToCSSPropMapping,
} from '../src/attributes'

it('transforms correctly', () => {
  const jsx = `
    function Test() {
      return (
        <div>
          <div
            mt="14em"
            mb="28px"
            ml="2%"
            mr={-14}
            pt={100}
            pl="0px"
            paddingBottom={200}
            className={['class1', 'class2'].join(' ')}>
            should create atomic classes
          </div>
          <CustomComponent pt={100} paddingBottom={200} className="foo bar">
            should create atomic classes
          </CustomComponent>
          <IgnoredComponent pt={100} paddingBottom={200}>
            should not create atomic classes
          </IgnoredComponent>  
        </div>
      )
    }
  `

  const result = traverse(jsx, 'test.tsx', {
    CSSPropToClassNameMapping,
    componentPropToCSSPropMapping,
    components: ['div', 'CustomComponent'],
  })
  expect((result || {}).code).toMatchSnapshot()
})
