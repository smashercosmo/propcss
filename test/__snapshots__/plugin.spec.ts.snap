// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transforms correctly 1`] = `
"function Test() {
  return <div>
          <div className={[\\"mt14 mb28 ml2% mr-14 pt100 pb200\\", ['class1', 'class2'].join(' ')].join(\\" \\")}>
            should create atomic classes
          </div>
          <CustomComponent className={[\\"pt100 pb200\\", \\"foo bar\\"].join(\\" \\")}>
            should create atomic classes
          </CustomComponent>
          <IgnoredComponent pt={100} paddingBottom={200}>
            should not create atomic classes
          </IgnoredComponent>  
        </div>;
}"
`;
